File: app/src/main/java/wxm/KeepAccount/Base/utility/PreferencesUtil.java
package wxm.KeepAccount.Base.utility;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Resources;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;

import wxm.KeepAccount.R;

/**
 * 配置管理类
 * Created by 123 on 2016/6/18.
 */
public class PreferencesUtil {
    private final static String PROPERTIES_NAME = "keepaccount_properties";
    private final static String SET_HOT_ACTION  = "hot_action";
    private final static String SET_CHART_COLOR = "chart_color";

    public final static String SET_PAY_COLOR    = "pay";
    public final static String SET_INCOME_COLOR = "income";
    public final static String SET_BUDGET_UESED_COLOR       = "budget_used";
    public final static String SET_BUDGET_BALANCE_COLOR     = "budget_balance";

    /// BEGIN
    /**
     * 加载首页上动作配置
     * @return   动作配置
     */
    public static List<String> loadHotAction()     {
        SharedPreferences param = ContextUtil.getInstance()
                                    .getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE);

        String def = String.format(Locale.CHINA,
                        "%s:%s:%s:%s:%s:%s",
                        ActionHelper.ACT_ADD_DATA, ActionHelper.ACT_LOOK_DATA,
                        ActionHelper.ACT_CALENDAR_VIEW, ActionHelper.ACT_LOOK_BUDGET,
                        ActionHelper.ACT_ADD_BUDGET, ActionHelper.ACT_LOGOUT);

        String load = param.getString(SET_HOT_ACTION, def);
        return parsePreferences(load);
    }


    /**
     * 保存首页上动作
     * @param acts 动作
     */
    public static void saveHotAction(List<String> acts)     {
        SharedPreferences param = ContextUtil.getInstance()
                .getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE);

        String pr = parseToPreferences(acts);
        param.edit().putString(SET_HOT_ACTION, pr).apply();
    }
    /// END


    /// BEGIN
    /**
     * 加载chart颜色配置
     * @return   颜色配置
     */
    public static HashMap<String, Integer> loadChartColor()     {
        SharedPreferences param = ContextUtil.getInstance()
                .getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE);

        Resources res = ContextUtil.getInstance().getResources();
        String sb = SET_PAY_COLOR + ":" + String.valueOf(res.getColor(R.color.sienna)) +
                " " + SET_INCOME_COLOR + ":" + String.valueOf(res.getColor(R.color.teal)) +
                " " + SET_BUDGET_UESED_COLOR + ":" + String.valueOf(res.getColor(R.color.sienna)) +
                " " + SET_BUDGET_BALANCE_COLOR + ":" + String.valueOf(res.getColor(R.color.teal));

        String load = param.getString(SET_CHART_COLOR, sb);
        return parseChartColors(load);
    }


    /**
     * 保存chart颜色配置
     * @param ccs  配色配置
     */
    public static void saveChartColor(HashMap<String, Integer> ccs)     {
        SharedPreferences param = ContextUtil.getInstance()
                .getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE);

        String pr = parseChartColorsToString(ccs);
        param.edit().putString(SET_CHART_COLOR, pr).apply();
    }
    /// END


    /// BEGIN PRIVATE
    /**
     * 把配置字符串解析成结果
     * @param pr        配置字符串
     * @return          结果
     */
    private static List<String> parsePreferences(String pr)  {
        ArrayList<String> ret = new ArrayList<>();
        String[] pr_arr = pr.split(":");
        Collections.addAll(ret, pr_arr);

        return ret;
    }

    /**
     * 把动作解析成配置字符串
     * @param acts      待解析动作
     * @return          配置字符串
     */
    private static String parseToPreferences(List<String> acts)  {
        boolean ff = true;
        StringBuilder sb = new StringBuilder();
        for(String i : acts)    {
            if(!ff) {
                sb.append(":");
            }   else    {
                ff = false;
            }

            sb.append(i);
        }

        return sb.toString();
    }

    /**
     * 从配置字符串解析配置
     * @param cc    配置字符串
     * @return   解析配置
     */
    private static HashMap<String, Integer> parseChartColors(String cc)  {
        HashMap<String, Integer> ret = new HashMap<>();
        String[] cclns = cc.split(" ");
        for(String i : cclns)   {
            String[] iilns = i.split(":");

            ret.put(iilns[0], Integer.parseInt(iilns[1]));
        }

        return ret;
    }


    private static String parseChartColorsToString(HashMap<String, Integer> hmcc)  {
        StringBuilder ret = new StringBuilder();
        for(String i : hmcc.keySet())   {
            StringBuilder sb = new StringBuilder();
            sb.append(i).append(":").append(hmcc.get(i).toString());

            if(0 == ret.length())
                ret.append(sb);
            else
                ret.append(" ").append(sb);
        }

        return ret.toString();
    }
    /// END PRIVATE
}

File: app/src/main/java/wxm/KeepAccount/preference/PreferencesUtil.kt
package wxm.KeepAccount.preference

import android.content.Context
import android.content.SharedPreferences
import org.greenrobot.eventbus.EventBus
import wxm.KeepAccount.R
import wxm.KeepAccount.define.EAction
import wxm.KeepAccount.event.PreferenceChange
import wxm.KeepAccount.utility.AppUtil
import java.util.*
import kotlin.collections.HashMap

/**
 * preference helper
 * Created by WangXM on 2016/6/18.
 */
class PreferencesUtil : PreferenceBase() {
    override fun loadPreference(pn:String, pd:String):String {
        return getPreferences().getString(pn, pd)
    }

    override fun savePreference(pn:String, pv:String)    {
        getPreferences().edit().putString(pn, pv).apply()
    }


    /// BEGIN PRIVATE
    private fun getPreferences(): SharedPreferences =
            AppUtil.self.getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE)
    /// END PRIVATE

    companion object {
        private const val PROPERTIES_NAME = "keep_account_properties"
        private const val DIVIDER_CHAR = ":"
        private const val DIVIDER_CHAR_1 = " "

        const val SET_PAY_COLOR = "pay"
        const val SET_INCOME_COLOR = "income"
        const val SET_BUDGET_UESED_COLOR = "budget_used"
        const val SET_BUDGET_BALANCE_COLOR = "budget_balance"

        private const val SET_HOT_ACTION = "hot_action"
        private const val SET_CHART_COLOR = "chart_color"

        private val preferenceObj = PreferencesUtil()

        /// BEGIN
        /**
         * load first page action setting
         * @return      action in first page
         */
        fun loadHotAction(): Array<String> {
            return String.format(Locale.CHINA,
                    "%s$DIVIDER_CHAR%s$DIVIDER_CHAR%s" +
                            "$DIVIDER_CHAR%s$DIVIDER_CHAR%s$DIVIDER_CHAR%s" +
                            "$DIVIDER_CHAR%s$DIVIDER_CHAR%s",
                    EAction.ADD_DATA.actName, EAction.LOOK_DATA.actName, EAction.CALENDAR_VIEW.actName,
                    EAction.LOOK_BUDGET.actName, EAction.ADD_BUDGET.actName, EAction.SYNC_SMS.actName,
                    EAction.ADD_DEBT.actName, EAction.LOGOUT.actName).let {
                preferenceObj.loadPreference(SET_HOT_ACTION, it)
            }.let {
                preferenceObj.parsePreference(it, DIVIDER_CHAR)
            }
        }

        /**
         * save fist page action
         * @param acts  action in fist page
         */
        fun saveHotAction(acts: List<String>) {
            preferenceObj.savePreference(SET_HOT_ACTION,
                    preferenceObj.parseToPreferences(acts, DIVIDER_CHAR))
            EventBus.getDefault().post(PreferenceChange(SET_HOT_ACTION))
        }
        /// END


        /// BEGIN
        /**
         * load chart color setting
         * @return  color setting
         */
        fun loadChartColor(): HashMap<String, Int> {
            return AppUtil.self.let {
                "$SET_PAY_COLOR:${it.getColor(R.color.sienna)}" +
                        " $SET_INCOME_COLOR:${it.getColor(R.color.teal)}" +
                        " $SET_BUDGET_UESED_COLOR:${it.getColor(R.color.sienna)}" +
                        " $SET_BUDGET_BALANCE_COLOR:${it.getColor(R.color.teal)}"
            }.let {
                preferenceObj.loadPreference(SET_CHART_COLOR, it)
            }.let {
                val ret = HashMap<String, Int>()
                preferenceObj.parsePreference(it, DIVIDER_CHAR_1, DIVIDER_CHAR)
                        .forEach {
                            ret[it.key] = Integer.parseInt(it.value)
                        }

                ret
            }
        }

        /**
         * save chart color setting
         * @param ccs   color setting
         */
        fun saveChartColor(ccs: HashMap<String, Int>) {
            preferenceObj.savePreference(SET_HOT_ACTION,
                    preferenceObj.parseToPreferences(ccs, DIVIDER_CHAR_1,  DIVIDER_CHAR))
            EventBus.getDefault().post(PreferenceChange(SET_CHART_COLOR))
        }
        /// END
    }
}
